crates/next-core/src/next_client_reference/visit_client_reference.rs

```diff
@@ -74,13 +95,31 @@ impl ClientReferenceGraphResult {
     }
 }
 
+impl ClientReferenceGraphResult {
+    /// Merges multiple return values of client_reference_graph together.
+    pub fn extend(&mut self, other: &Self) {
+        self.client_references
+            .extend(other.client_references.iter().copied());
+        for (k, v) in other.client_references_by_server_component.iter() {
+            self.client_references_by_server_component
+                .entry(*k)
+                .or_insert_with(Vec::new)
+                .extend(v);
+        }
+        self.server_component_entries
+            .extend(other.server_component_entries.iter().copied());
+        self.server_utils.extend(other.server_utils.iter().copied());
+        // This is merged already by `client_reference_graph` itself
+        self.visited_nodes = other.visited_nodes;
+    }
+}
+
 #[turbo_tasks::function]
 pub async fn client_reference_graph(
-    entries: Vc<Modules>,
+    entry: Vc<Box<dyn Module>>,
+    visited_nodes: Vc<VisitedClientReferenceGraphNodes>,
 ) -> Result<Vc<ClientReferenceGraphResult>> {
```

Comment:
You could split that into two graph walks.

1. walking from the RSC entry to ServerComponents and ServerUtils (stop when these are found)
2. walking from each ServerComponent and ServerUtil (with visited modules) to ClientComponents

`1.` is pretty quick
`2.` can be optimized via visited modules/layout segment optimization

Reply:
Done. This has the additional benefit now that it also caches processes the `ServerUtils` just once. That was still done repeatedly in my first version